code/computational_geometry/src/sphere_tetrahedron_intersection/sphere_tetrahedron_intersection.cpp

```diff
@@ -0,0 +1,1237 @@
+		}
+		else if( radius > lenA )
+		{
+			return solidAngleFull
+				- SolidAngleCap( angERA ) * angBAC / ( 2.0 * pi );
+		}
+		else
+		{
+			return solidAngleFull;
+		}
+	}
+	
+	double GetSurfaceArea( double radius )
+	{
+		return GetSolidAngle( radius ) * radius * radius;
+	}
+	
+	double GetVolume( double radius )
+	{
+		Recalculate( radius );
+		if( radius > lenC )
+		{
+			return VolumeTetrahedron( posR, posA, posB, posC );
+		}
+		else if( radius > lenB )
+		{
+			return ( VolumeWedge( dirA, dirC, dirF, radius ) - (
+				VolumeCap( radius, LUV::Length( prpA - posA ) )
+				- VolumeCone( prlE, lenA )
+				) * angFAE / ( 2.0 * pi ) )
+				+ VolumeTetrahedron( posR, posA, posB, prpF );
+		}
+		else if( radius > lenA )
+		{
+			return VolumeWedge( dirA, dirB, dirC, radius ) - (
+				VolumeCap( radius, LUV::Length( prpA - posA ) )
+				- VolumeCone( prlE, lenA )
+				) * angBAC / ( 2.0 * pi );
+		}
+		else
+		{
+			return VolumeWedge( dirA, dirB, dirC, radius );
+		}
+	}
+	
+};
+
+class SphereTetrahedronIntersection
+{
+	
+private:
+
+	std::vector< ObservationTetrahedron > obsTet;
+	std::vector< double > obsTetSgn;
+	std::vector< Vec3 > dirN;
+	std::vector< Vec3 > dirU;
+	std::vector< Vec3 > dirL;
+	
+	void AddEdge( const Vec3& vecM, const Vec3& vecP, const Vec3& vecT, const Vec3& vecR )
+	{
+		Vec3 PmM = vecP - vecM;
+		dirL.push_back( PmM / LUV::Length( PmM ) );
+		Vec3 vecU = vecT - LUV::ProjLine( vecT, vecM, vecP );
+		dirU.push_back( vecU / LUV::Length( vecU ) );
+		Vec3 vecA = LUV::ProjPlane( vecR, vecM, LUV::PlaneNormal( vecM, vecP, vecT ) );
+		Vec3 vecB = LUV::ProjLine( vecA, vecM, vecP );
+		obsTet.push_back( ObservationTetrahedron( vecR, vecA, vecB, vecP ) );
+		obsTet.push_back( ObservationTetrahedron( vecR, vecA, vecB, vecM ) );
+	}
+	
+	void AddFace( const Vec3& vecA, const Vec3& vecB, const Vec3& vecC, const Vec3& vecU , const Vec3& vecR )
+	{
+		Vec3 vecN = vecU - LUV::ProjPlane( vecU, vecA, LUV::PlaneNormal( vecA, vecB, vecC ) );
+		dirN.push_back( vecN / LUV::Length( vecN ) );
+		AddEdge( vecA, vecB, vecC, vecR );
+		AddEdge( vecB, vecC, vecA, vecR );
+		AddEdge( vecC, vecA, vecB, vecR );
+	}
+	
+public:
+	
+	SphereTetrahedronIntersection( const Vec3& vecTetA, const Vec3& vecTetB, const Vec3& vecTetC, const Vec3& vecTetD, const Vec3& vecSphCenter )
+	{
+		AddFace( vecTetA, vecTetB, vecTetC, vecTetD, vecSphCenter );
+		AddFace( vecTetB, vecTetC, vecTetD, vecTetA, vecSphCenter );
+		AddFace( vecTetC, vecTetD, vecTetA, vecTetB, vecSphCenter );
+		AddFace( vecTetD, vecTetA, vecTetB, vecTetC, vecSphCenter );
+		for( int idf = 0; idf < 4; ++idf )
+		{
+			for( int idl = 0; idl < 3; ++idl )
+			{
+				int ids = 3 * idf + idl;
+				int idp = 2 * ids;
+				int idm = idp + 1;
+				
+				obsTetSgn.push_back(
+					LUV::Dot( obsTet[idp].prdA, dirN[idf] )
+					* LUV::Dot( obsTet[idp].prdF, dirL[ids] )
+					* LUV::Dot( obsTet[idp].prdD, dirU[ids] )
+				);
+				
+				obsTetSgn.push_back(
+					- LUV::Dot( obsTet[idm].prdA, dirN[idf] )
```

Comment:
This looks more confusing. If you mean the unary minus operator, then it should be attached to its number. The operations should be on the right hand side, not the left.

Reply:
Yes unary.

"The operations should be on the right hand side, not the left."

You mean this:
- A
* B
* C;

must be this:
-A*
B*
C;

Reply:
No, I mean:

```
-A *
B *
C;
```